File: src/rod_cutting/rod_cutting.py
def rod_cutting(n, price, length):
    if n == 0:
        return 0
    value = 0
    for i in range(1, n + 1):
        newValue = price[i] + rod_cutting(n-i, price, length)
        if newValue > value:
            value = newValue
            length[n] = i
    
    return value

def rod_cutting_memo(n, price, memo, length):
    if memo[n] != -1:
        return memo[n]
    memo[n] = 0
    for i in range(1, n + 1):
        newValue = price[i] + rod_cutting_memo(n-i, price, memo, length)
        if newValue > memo[n]:
            memo[n] = newValue
            length[n] = i
    
    return memo[n]

def rod_cutting_iter(n, price, length):
    memo = [0]
    
    for i in range(1, n + 1):
        value = 0
        for j in range(1, i + 1):
            newValue = price[j] + memo[i-j]
            if newValue > value:
                value = newValue
                length[i] = j
        memo.append(value)
        
    return memo[n]

def printLength(n, length):
    while n > 0:
        print(length[n])
        n = n - length[n]

File: src/rod_cutting/rod_cutting.cpp
#include <stdio.h>

int rod_cutting(int n, int* product, int* length)
{
	if (n == 0)
	{
		return 0;
	}

	int res = 0;

	for (int i = 1; i <= n; i++)
	{
		int value = product[i] + rod_cutting(n - i, product);

		if (value > res)
		{
			res = value;
			length[n] = i;
		}
	}
	
	return res;
}

int rod_cutting_memo(int n, int* product, int* memo, int* length)
{
	if (memo[n] != -1)
		return memo[n];

	if (n == 0)
		memo[n] = 0;

	for (int i = 1; i <= n; i++)
	{
		int value = product[i] + rod_cutting_memo(n - i, product, memo);

		if (value > memo[n])
		{
			memo[n] = value;
			length[n] = i;
		}
	}

	return memo[n];
}

int rod_cutting_iter(int n, int* product, int* length)
{
	int* arr = new int[n + 1];

	arr[0] = 0;

	for (int i = 1; i <= n; i++)
	{
		arr[i] = 0;
		for (int j = 1; j <= i; j++)
		{
			int value = product[j] + arr[i - j];

			if (value > arr[i])
			{
				arr[i] = value;
				length[i] = j;
			}
		}
	}

	int res = arr[n];

	delete[] arr;

	return res;
}

void printLength(int n, int* length)
{
	while (n > 0)
	{
		printf("%d ", length[n]);
		n = n - length[n];
	}

	printf("\n");
}